src/analysis/processing/qgsalgorithmjoinbylocation.cpp

```diff
@@ -0,0 +1,368 @@
+/***************************************************************************
+                         qgsalgorithmjoinbylocation.cpp
+                         ---------------------
+    begin                : November 2019
+    copyright            : (C) 2019 by
+    email                :
+ ***************************************************************************/
+
+/***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsalgorithmjoinbylocation.h"
+#include "qgsprocessing.h"
+#include "qgsgeometryengine.h"
+#include "qgsvectorlayer.h"
+#include "qgsapplication.h"
+#include "qgsfeature.h"
+#include "qgsfeaturesource.h"
+
+///@cond PRIVATE
+
+
+void QgsJoinByLocationAlgorithm::initAlgorithm( const QVariantMap & )
+{
+  addParameter( new QgsProcessingParameterFeatureSource( QStringLiteral( "INPUT" ),
+                QObject::tr( "Base Layer" ), QList< int > () << QgsProcessing::QgsProcessing::TypeVectorAnyGeometry ) );
+  addParameter( new QgsProcessingParameterFeatureSource( QStringLiteral( "JOIN" ),
+                QObject::tr( "Join Layer" ), QList< int > () << QgsProcessing::QgsProcessing::TypeVectorAnyGeometry ) );
+
+  QStringList predicates;
+  predicates << QObject::tr( "intersects" )
+             << QObject::tr( "contains" )
+             << QObject::tr( "isEqual" )
+             << QObject::tr( "touches" )
+             << QObject::tr( "overlaps" )
+             << QObject::tr( "within" )
+             << QObject::tr( "crosses" );
+
+  addParameter( new QgsProcessingParameterEnum( QStringLiteral( "PREDICATE" ), QObject::tr( "Geometric predicate" ),
```

Comment:
```suggestion
  std::unique_ptr< QgsProcessingParameterEnum > predicateParam = qgis::make_unique< QgsProcessingParameterEnum >( QStringLiteral( "PREDICATE" ), QObject::tr( "Geometric predicate" ), predicates, true, 0 );
  QVariantMap predicateMetadata;
  QVariantMap widgetMetadata;
  widgetMetadata.insert( QStringLiteral( "useCheckBoxes" ), true );
  widgetMetadata.insert( QStringLiteral( "columns" ), 2 );
  predicateMetadata.insert( QStringLiteral( "widget_wrapper" ), widgetMetadata );
  predicateParam->setMetadata( predicateMetadata );
  addParameter( predicateParam.release() );
```
( yep, much more concise in the original Python version!!)